File: Algorithms/src/1979. Find Greatest Common Divisor of Array/Code/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */

// Function to find GCD using the Euclidean Algorithm
function gcd(a, b) {
  // Repeat until b becomes 0
  while (b !== 0) {
    let temp = b; // Store current b
    b = a % b; // Update b with remainder
    a = temp; // Move previous b into a
  }

  // a is the GCD
  return a;
}

var findGCD = function (nums) {
  // Initialize minimum and maximum
  let min = nums[0];
  let max = nums[0];

  // Find smallest and largest values
  for (const num of nums) {
    min = Math.min(min, num);
    max = Math.max(max, num);
  }

  // Return their GCD
  return gcd(min, max);
};

File: Algorithms/src/1979. Find Greatest Common Divisor of Array/Code/solution.go
// Function to find GCD using the Euclidean Algorithm
func gcd(a, b int) int {
	// Continue until b becomes 0
	for b != 0 {
		a, b = b, a%b
	}

	// a now stores the GCD
	return a
}

func findGCD(nums []int) int {

	// Initialize minimum and maximum with the first element
	minimum := nums[0]
	maximum := nums[0]

	// Find the smallest and largest values
	for _, num := range nums {
		if num < minimum {
			minimum = num
		}
		if num > maximum {
			maximum = num
		}
	}

	// Return the GCD of the smallest and largest numbers
	return gcd(minimum, maximum)
}